### DISCUSSION TITLE
How to do client streaming with `Flux` argument to `HttpRequest`

### DISCUSSION THREAD

**Discussion by RonBarkan** [micronaut-projects/micronaut-core#10787] [Q&A]
It is required to "upload" a programmatically generated content. The content might be large an it is not desired to store the entire content in memory at any given time. I am trying to use a low-level HttpClient for this.
This is not documented [here](https://docs.micronaut.io/latest/guide/index.html#httpClient), [here](https://github.com/micronaut-projects/micronaut-core/blob/e9554c6a5ca0bb588eed60b77b08d601aa788c17/http/src/main/java/io/micronaut/http/HttpRequest.java#L34) nor [here](https://guides.micronaut.io/latest/micronaut-http-client-gradle-java.html), but I was able to find some hints ([1](https://github.com/micronaut-projects/micronaut-core/blob/4.5.x/http-client-tck/src/main/java/io/micronaut/http/client/tck/tests/ContinueTest.java#L69), [2](https://github.com/micronaut-projects/micronaut-core/issues/2906)) that this might be allowed.

I have some code that sends JSON requests to some server and I've tweaked it to do the following:

```
   // var req = HttpRequest.POST(uri, jsonString);          // Original line (this works!)
   var req = HttpRequest.POST(uri, Mono.just(jsonString));  // Replacement (does not work)
    req.contentType(contentType)
        .accept(MediaType.APPLICATION_JSON)
        .bearerAuth(context.getAccessToken());

    return Mono.from(client.exchange(req, String.class));
```

When I run the code against a real server, I see that:
* POST header is sent and after a long deadline, I get 500 from the server (in the case of a long deadline), or timeout exception (in the case of the default client config)
* The Mono I introduced is not subscribed to by anything (I have added a `.doOnSubscribe(...)` that is never called)
* Tracing shows that the POST header is being sent with `Transfer-Encoding: chunked`, but then there's no more traces of the body being sent, so maybe it is never sent (tracing is done with ` <logger name="io.micronaut.http.client" level="TRACE"/>`)
* I have used `curl` to send an identical message to the server, chunked encoded, and I get success

Because of all the magic and the abstractions, it is really not clear what's wrong. I expected `Mono.just(jsonString)` to work without issue.
From the documentation, it's unclear what is the relationship between `Flux`, `Mono` and content type.

Can you please help explain what might be wrong here, especially if the body is never sent to the server.
Note that the server and protocol cannot be changed. `Mono.just()` is used here only as a first step, to see that it works (it does not).
Using Micronaut v3.10.1

**Comment by RonBarkan** [comment]
@yawkat 
@timyates 
@graemerocher 
Any suggestions?

**Reply by graemerocher** [reply]
Use `jsonStream` method

**Reply by graemerocher** [reply]
not sure if at this stage we support streaming the request body, maybe @yawkat knows

**Reply by yawkat** [reply]
publisher should work, not sure why it doesnt here.

**Reply by RonBarkan** [reply]
The issue materializes in a real server, which acts as a client to some 3rd party server, and uses a "low-level HttpClient".
Here is what I have tried:

1. I wrote an integration test, which shows `Flux/Mono` upload works. Unfortunately, this test works, so I can't use it to file an issue.
2. Enabled HttpClient tracing with `<logger name="io.micronaut.http.client" level="TRACE"/>`. Unless tracing itself has a bug, HttpClient never sends the request body after the headers.
3. As mentioned above, I've added some logging to the body `Mono`:  `Mono.just(jsonString).doOnSubscribe(s -> log.info("Subscribed!"))`, which is then passed to the `POST()` method (see above). This never logs, which means it cannot possibly be consumed.
4. I've reset the `HttpClient` configuration to default. It had no impact.
5. I have started to step through Micronaut code, and it does look like processing paths are different between (1) and my server trying to do the same. However, it will take me a long time to explore this, since I have no idea what to look for, other than compare the processing path of a good and bad examples.

Any suggestions?

**Reply by graemerocher** [reply]
provide an example in an issue report and we can have a look
